script/tool/lib/src/common/package_looping_command.dart

```diff
@@ -22,6 +22,18 @@ abstract class PackageLoopingCommand extends PluginCommand {
     GitDir? gitDir,
   }) : super(packagesDir, processRunner: processRunner, gitDir: gitDir);
 
+  /// Packages that had a [logSkip] call.
+  Set<Directory> _skippedPackages = <Directory>{};
+
+  /// Packages that had at least one [logWarning] call.
+  Set<Directory> _packagesWithWarnings = <Directory>{};
+
+  /// Number of warnings that happened outside of a [runForPackage] call.
+  int _otherWarningCount = 0;
+
+  /// The package currently being run by [runForPackage].
+  Directory? _currentPackage;
```

Comment:
> It would be better off as a parameter to logSkip and logWarning.

I strongly considered that, since this approach feels a bit sketchy, but in practice that would be worse. With this version, the possible bugs are contained entirely within this one base class, which is short and has unit tests ensuring that these work correctly. With the parameter version, however, it's possible to write things like:

`logWarning(message, example)` // Oops, this was supposed to be `package`, not the `example` subdirectory.

and thus causing the warning not be summarized anywhere.

For `logSkip`, the more I think about the summary, the more I think I should actually make it a return value option instead. Originally I had created printSkip for logging any case where something is being skipped, which for some commands might be, say, a subset of the examples. But with the summary we don't actually want to allow that, so I need to rework the return value. I'll land out the outstanding other conversions, then change the value to a return class with state (pass/skip/error)+error messages.

> Keeping it around could lead to some weird bugs at some point because the coupling between logSkip and _currentPackage isn't obvious to the client.

For `logSkip` I agree. For `logWarning` I think that's actually a feature; a command can just logging warnings whenever it wants to as output, and doesn't have to care about what the base class might be doing with them. How they are tracked and summarized is entirely an implementation detail of the base.

Reply:
This PR has been updated to eliminate `logSkip` by changing the return from a run to a small data class and making "skipped" a first-class result rather than a possibly-partial bit based on a log.

Commands that used to log formal skips for subsets of the command while running other subsets now just use simple print statements to indicate what is and isn't being run, and "skip" is reserved for cases where nothing is run.

Reply:
This is a huge improvement.  You could eliminate `_currentPackage` if you created a way for PackageResult to carry the warnings too.  It's not a huge deal, `_currentPackage` bugs me a bit since it's a looping parameter in an instance variable.

Reply:
It's valid to print warnings outside the loop though; I don't want to have two different ways of handling warnings in the subclasses.